File: Data Structures & Algorithms/trapping-rain-water/submission-1.cpp
// class Solution {
//     public int trap(int[] height) {
//         Stack<Integer> st = new Stack<Integer>();
//         int area = 0;
//         for(int i = 0;i<height.length;i++){
//             int curr = height[i];
//             if(st.empty()){
//                 st.push(i);
//             } 
//             else{
//                 int base = height[st.peek()];
//                 if(curr < base) st.push(i);
//                 else{
//                     while(curr > base){
//                         st.pop();
//                         if(st.empty()) break;
//                         int top = st.peek();
//                         int ht = Math.min(height[top], curr) - base;
//                         int width = i - top - 1;
//                         area = area + (ht * width);
//                         base = height[top];
//                     }
//                     st.push(i);
//                 }
//             }
//         }
//         return area;
//     }
// }
class Solution {
public:
    int trap(vector<int>& height) {
        int left = 0;
        int right = height.size() - 1;
        int leftMax = -1;
        int rightMax = -1;
        int ans = 0;
        while(left < right){
            leftMax = max(leftMax, height[left]);
            rightMax = max(rightMax, height[right]);

            if(leftMax < rightMax){
                ans = ans + (leftMax - height[left]);
                left++;
            }
            else{
                ans = ans + (rightMax - height[right]);
                right--;
            }
        }
        return ans;
    }
};

File: Data Structures & Algorithms/min-cost-climbing-stairs/submission-1.java
class Solution {
    private int f(int ind, int[] cost, int n, int[] dp){
        // base
        if(ind >= n) return 0;
        if(dp[ind]!=-1) return dp[ind];
        int one = f(ind + 1, cost, n, dp);
        int two = f(ind + 2, cost, n, dp);
        return dp[ind] = cost[ind] + Math.min(one, two);
    }
    public int minCostClimbingStairs(int[] cost) {
        int n = cost.length;
        int[] dp = new int[n + 2];
        // for(int i = 0; i < n; i++) dp[i] = -1;
        // dp[n] = 0;
        // dp[n + 1] = 0;
        int prev1 = 0, prev2 = 0, curr;
        for(int ind = n - 1; ind >=0; ind--){
            // int one = dp[ind + 1];
            // int two = dp[ind + 2];
            curr = cost[ind] + Math.min(prev1, prev2);
            prev2 = prev1;
            prev1 = curr;
        }
        return Math.min(prev1, prev2);
    }
}